TASK: Adjust document uri projection to soft removals

mhsdesign opened the pull request:
Soft removals were introduced via https://github.com/neos/neos-development-collection/pull/5479.
They work by tagging a node removed instead of removing the node directly. This is just a Neos convention ontop of the CR.
Now the document uri path projection needs to be aware of that, as otherwise it will attempt resolve uris when the node is already removed.

This is archived by using the same ideas of the already implemented "disabled" state handling, we just extract some common logic and introduce a new column "removed" next to "disabled". It also uses an integer to indicate if a node was hidden or if it was hidden though its parent.

Initially i hoped to be able to get rid of the `isNodeExplicitlyDisabled` functionality in all places as this check has become mostly redundant. Its currently only required to project node moves correctly, as there disabled nodes must stay so if moved.

Now the `isNodeExplicitlyDisabled` check was dropped as discussed for the case of just updating the disabled state. 
Previously the projection could handle a case where a node was disabled, or re-enabled after already being in that state.
While partly controversial, we agreed that this behaviour is not valid and is to be caught by constraint checks as implemented by https://github.com/neos/neos-development-collection/pull/5357
If however we decide to allow to disable a node twice - the same as setting a property to the same value again - we must adjust the projection to be robust for both the disabled and removed state, and then introduce test using the actual command api instead of hackily publishing events.


**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
TASK: Adjust document uri projection to soft removals

mhsdesign committed:
TASK: Do not attempt to handle wrong events (e.g.) a node being disabled twice

With https://github.com/neos/neos-development-collection/pull/5357 this CANNOT happen, this test is just a leftover and isNodeExplicitlyDisabled is not a necessity anymore.

mhsdesign commented on Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathFinder.php:
```diff
@@ -80,6 +81,7 @@ public function getByIdAndDimensionSpacePointHash(
         if ($this->cacheEnabled && isset($this->getByIdAndDimensionSpacePointHashCache[$cacheKey])) {
             return $this->getByIdAndDimensionSpacePointHashCache[$cacheKey];
         }
+        // todo AND removed = 0 -> we should not allow building uris to removed nodes!
```
This has to be discussed!

Currently matching and resolving is not behaving the same for disabled nodes:

As shown by this test:

```
Then No node should match URL "/david-nodenborough/earl-document/leaf"
And The node "leaf-mc-node" in content stream "cs-identifier" and dimension "{}" should resolve to URL "/david-nodenborough/earl-document/leaf"
```

so matching is expected to work always, but i think this is wrong! And i dont want to introduce this for soft removed nodes.

kitsunet replied:
it would be utterly broken if we introduce it for soft deleted nodes. it must free the url.

kitsunet replied:
Oh look:

> Note: This will not exclude *disabled* nodes in order to allow the calling side to make a distinction (e.g. in order to display a custom error)

This is from the docblock at `\Neos\Neos\FrontendRouting\Projection\DocumentUriPathFinder::getByIdAndDimensionSpacePointHash`

kitsunet replied:
Which for example the shortcut resolver uses to throw a custom error about the target being disabled.

I guess we need split behavior where disabled does resolve and removed doesnt'

mhsdesign replied:
or we just dont do that for now as this introduces just additional complexity. There are a lot of cases where disabled nodes cannot be worked with like when using back references or references but the nodes are hidden and thus wont be displayed in fusion live rendering. My point is, its fairly obvious that linking should not work and editors who want to find out why not will not have access to the stack trace and have to find out either way by looking at the tree, and why make it complex just to make it easy for developers.
I think we should have a clean api to the outside here.

mhsdesign replied:
okay found it the resolving (not matching!) of disabled nodes was done via https://github.com/neos/neos-development-collection/pull/4363 to make the redirect handler work.

So we sadly cannot remove the complexity to return disabled nodes and have to split decisions for soft removed nodes.

mhsdesign committed:
LOL: Fix `isNodeExplicitlyRemoved` implementation

mhsdesign committed:
TASK: Disallow uri building to soft removed nodes

Contraire resolving disabled nodes: https://github.com/neos/neos-development-collection/pull/4363/
We forbid that for removed nodes.

mhsdesign committed:
TASK: Add test that `getEnabledBySiteNodeNameUriPathAndDimensionSpacePointHash` filters nodes by disabled and removed state

they must not be matched against a live uri!

mhsdesign committed:
TASK: Remove obsolet `in content stream "cs-identifier"` from frontend uri projection tests

This is a leftover from the workspace aware node adjustments

mhsdesign committed:
TASK: Declare `DocumentUriPathFinder` & co as internal

The method `getByIdAndDimensionSpacePointHash` will for example now return soft removed `DocumentNodeInfo`'s to make this a stable api this needs to get an overhaul.

Neos takes already care to handle all tasks like shortcut resolving or uri matching. In case certain things need to be made we should provide them as a service to be able to swap out the underlying implementation and to not expose things like `DocumentNodeInfos`.

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into task/adjust-document-uri-projection-to-soft-removals

mhsdesign committed:
TASK: Use new `NeosSubtreeTag::removed()` instead of temporary stub

as per https://github.com/neos/neos-development-collection/pull/5463

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into task/adjust-document-uri-projection-to-soft-removals

mhsdesign committed:
TASK: Adjust to `SubtreeTagging` change namespace

kitsunet reviewed (approved):
Thank you
